# University of Michigan 
# Master of Applied Data Science
### SIAD 699: Capstone
Team: James Yoon (jamyoon), Mario Feliciano (felicma), and James Tuccori (jtuccori)

Date: April 2023
____
# Social Media Sentiment & Predicting Trading Signals 

#### The purpose of this notebook is to assess all the different paramters for the LSTM model to be able to see the best paramters to select. Our assessment and logic for selecting Timestep: 12 ; Neurons: 128; Dropout: 0.1,  Learning: 0.001; With a difference of 0.001687132, LSTM Acc. AVG of 0.530923147, and Val Acc.AVG of 0.529236015. 

In [1]:
import pandas as pd
from scipy import stats
import statsmodels.formula.api as smf
from stargazer.stargazer import Stargazer

In [49]:

# Load the data
a1 = pd.read_csv('LSTM Paramter Assessment/AnalysisConsolidates testingresultsJTGraph.csv')

In [50]:
a1.columns

Index(['Unnamed: 0', 'LSTM_Acc_AVG', 'Val_Acc_AVG', 'Timestep', 'Neurons',
       'Dropout', 'Learning', 'difference'],
      dtype='object')

In [51]:
a1 = a1[['LSTM_Acc_AVG', 'Val_Acc_AVG', 'Timestep', 'Neurons',
       'Dropout', 'Learning', 'difference']]
a1

LSTM_Acc_AVG  Val_Acc_AVG  Timestep  Neurons  Dropout  Learning  \
0        0.520318     0.514168         2        2      0.1     0.010   
1        0.529273     0.522300         2        2      0.1     0.001   
2        0.517909     0.512745         2        2      0.2     0.010   
3        0.528545     0.520268         2        2      0.2     0.001   
4        0.520773     0.515682         2        2      0.3     0.010   
..            ...          ...       ...      ...      ...       ...   
220      0.525217     0.526243       168      128      0.2     0.001   
221      0.517208     0.508984       168      128      0.3     0.010   
222      0.524989     0.524549       168      128      0.3     0.001   
223      0.505767     0.507336       168      128      0.4     0.010   
224      0.526911     0.527057       168      128      0.4     0.001   

     difference  
0      0.006150  
1      0.006973  
2      0.005164  
3      0.008277  
4      0.005091  
..          ...  
220   -0.001025  
221    0.008224  
222    0.000439  
223   -0.001570  
224   -0.000146  

[225 rows x 7 columns]

In [21]:
a1.columns

Index(['LSTM_Acc_AVG', 'Val_Acc_AVG', 'Timestep', 'Neurons', 'Dropout',
       'Learning', 'difference'],
      dtype='object')

In [25]:
m1 = smf.ols(formula = 'LSTM_Acc_AVG ~Timestep',  data = a1).fit()
reg_q2_1 = m1
model_summary= reg_q2_1.summary()
a1_q2_ate = reg_q2_1.params[1]
print(a1_q2_ate)
print(model_summary)

-1.506023033885689e-05
                            OLS Regression Results                            
Dep. Variable:           LSTM_Acc_AVG   R-squared:                       0.016
Model:                            OLS   Adj. R-squared:                  0.012
Method:                 Least Squares   F-statistic:                     3.739
Date:                Sun, 19 Mar 2023   Prob (F-statistic):             0.0544
Time:                        13:30:51   Log-Likelihood:                 792.73
No. Observations:                 225   AIC:                            -1581.
Df Residuals:                     223   BIC:                            -1575.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.5260      0.

In [24]:
m1 = smf.ols(formula = 'LSTM_Acc_AVG ~Timestep + Neurons + Dropout + Learning ',  data = a1).fit()
reg_q2_1 = m1
model_summary= reg_q2_1.summary()
a1_q2_ate = reg_q2_1.params[1]
print(a1_q2_ate)
print(model_summary)

-1.5431066461557058e-05
                            OLS Regression Results                            
Dep. Variable:           LSTM_Acc_AVG   R-squared:                       0.279
Model:                            OLS   Adj. R-squared:                  0.266
Method:                 Least Squares   F-statistic:                     21.27
Date:                Sun, 19 Mar 2023   Prob (F-statistic):           7.62e-15
Time:                        13:30:47   Log-Likelihood:                 827.64
No. Observations:                 225   AIC:                            -1645.
Df Residuals:                     220   BIC:                            -1628.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.5346      0

In [26]:

#Creating Dummy Values(columns) to see which timestep value is significant
df = pd.get_dummies(a1, columns=['Timestep'], prefix='Timestep')
df

LSTM_Acc_AVG  Val_Acc_AVG  Neurons  Dropout  Learning  difference  \
0        0.520318     0.514168        2      0.1     0.010    0.006150   
1        0.529273     0.522300        2      0.1     0.001    0.006973   
2        0.517909     0.512745        2      0.2     0.010    0.005164   
3        0.528545     0.520268        2      0.2     0.001    0.008277   
4        0.520773     0.515682        2      0.3     0.010    0.005091   
..            ...          ...      ...      ...       ...         ...   
220      0.525217     0.526243      128      0.2     0.001   -0.001025   
221      0.517208     0.508984      128      0.3     0.010    0.008224   
222      0.524989     0.524549      128      0.3     0.001    0.000439   
223      0.505767     0.507336      128      0.4     0.010   -0.001570   
224      0.526911     0.527057      128      0.4     0.001   -0.000146   

     Timestep_2  Timestep_6  Timestep_12  Timestep_48  Timestep_96  \
0             1           0            0            0            0   
1             1           0            0            0            0   
2             1           0            0            0            0   
3             1           0            0            0            0   
4             1           0            0            0            0   
..          ...         ...          ...          ...          ...   
220           0           0            0            0            0   
221           0           0            0            0            0   
222           0           0            0            0            0   
223           0           0            0            0            0   
224           0           0            0            0            0   

     Timestep_168  
0               0  
1               0  
2               0  
3               0  
4               0  
..            ...  
220             1  
221             1  
222             1  
223             1  
224             1  

[225 rows x 12 columns]

In [28]:
m1 = smf.ols(formula = 'LSTM_Acc_AVG ~ Timestep_2 + Timestep_6 + Timestep_12 + Timestep_48 + Timestep_96 + Timestep_168',  data = df).fit()
reg_q2_1 = m1
model_summary= reg_q2_1.summary()
a1_q2_ate = reg_q2_1.params[1]
print(a1_q2_ate)
print(model_summary)

0.07584329618723859
                            OLS Regression Results                            
Dep. Variable:           LSTM_Acc_AVG   R-squared:                       0.020
Model:                            OLS   Adj. R-squared:                 -0.003
Method:                 Least Squares   F-statistic:                    0.8712
Date:                Sun, 19 Mar 2023   Prob (F-statistic):              0.501
Time:                        13:34:39   Log-Likelihood:                 793.07
No. Observations:                 225   AIC:                            -1574.
Df Residuals:                     219   BIC:                            -1554.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        0.4502     

In [ ]:
# All values are significant 

In [31]:
#Grouping by the values by the mean values

a2 = a1.copy()
a2

LSTM_Acc_AVG  Val_Acc_AVG  Timestep  Neurons  Dropout  Learning  \
0        0.520318     0.514168         2        2      0.1     0.010   
1        0.529273     0.522300         2        2      0.1     0.001   
2        0.517909     0.512745         2        2      0.2     0.010   
3        0.528545     0.520268         2        2      0.2     0.001   
4        0.520773     0.515682         2        2      0.3     0.010   
..            ...          ...       ...      ...      ...       ...   
220      0.525217     0.526243       168      128      0.2     0.001   
221      0.517208     0.508984       168      128      0.3     0.010   
222      0.524989     0.524549       168      128      0.3     0.001   
223      0.505767     0.507336       168      128      0.4     0.010   
224      0.526911     0.527057       168      128      0.4     0.001   

     difference  
0      0.006150  
1      0.006973  
2      0.005164  
3      0.008277  
4      0.005091  
..          ...  
220   -0.001025  
221    0.008224  
222    0.000439  
223   -0.001570  
224   -0.000146  

[225 rows x 7 columns]

In [33]:
a2.groupby('Timestep').mean()

LSTM_Acc_AVG  Val_Acc_AVG    Neurons   Dropout  Learning  difference
Timestep                                                                      
2             0.526040     0.520271  47.000000  0.250000   0.00550    0.005769
6             0.525515     0.520008  47.000000  0.250000   0.00550    0.005507
12            0.526616     0.521006  58.250000  0.250000   0.00550    0.005610
48            0.525313     0.521021  58.250000  0.250000   0.00550    0.004292
96            0.524169     0.518698  47.439024  0.253659   0.00561    0.005471
168           0.523723     0.519104  47.000000  0.250000   0.00550    0.004619

#### To test if the LSTM_Acc_AVG values of each timestep are significantly different,
#### we perform an Analysis of Variance (ANOVA) using the statsmodels library.

In [35]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

# Replace 'data' with the name of your DataFrame
model = ols('LSTM_Acc_AVG ~ C(Timestep)', data=a2).fit()
anova_table = sm.stats.anova_lm(model, typ=2)
print(anova_table)

               sum_sq     df        F    PR(>F)
C(Timestep)  0.000227    5.0  0.87122  0.501193
Residual     0.011432  219.0      NaN       NaN


####  In this case, the p-value (PR(>F)) is 0.501193, which is greater than the common significance threshold of 0.05. This means that we fail to reject the null hypothesis, suggesting that there is no significant difference between the LSTM_Acc_AVG values of each timestep. In other words, the differencesin LSTM_Acc_AVG values across the timesteps may be due to random chance rather than being an  actual effect of the timesteps.

#### So to choose the best parameters for our model, we took the combination of hyperparameters that leads to the highest validation accuracy (Val_Acc_AVG) while maintaining a small difference between training and validation accuracy to avoid overfitting. 

In [52]:
a3 = a1.copy()
a3

LSTM_Acc_AVG  Val_Acc_AVG  Timestep  Neurons  Dropout  Learning  \
0        0.520318     0.514168         2        2      0.1     0.010   
1        0.529273     0.522300         2        2      0.1     0.001   
2        0.517909     0.512745         2        2      0.2     0.010   
3        0.528545     0.520268         2        2      0.2     0.001   
4        0.520773     0.515682         2        2      0.3     0.010   
..            ...          ...       ...      ...      ...       ...   
220      0.525217     0.526243       168      128      0.2     0.001   
221      0.517208     0.508984       168      128      0.3     0.010   
222      0.524989     0.524549       168      128      0.3     0.001   
223      0.505767     0.507336       168      128      0.4     0.010   
224      0.526911     0.527057       168      128      0.4     0.001   

     difference  
0      0.006150  
1      0.006973  
2      0.005164  
3      0.008277  
4      0.005091  
..          ...  
220   -0.001025  
221    0.008224  
222    0.000439  
223   -0.001570  
224   -0.000146  

[225 rows x 7 columns]

#### We sort the  data table by Val_Acc_AVG in descending order and then look for the combinations that have the highest Val_Acc_AVG along with a relatively low difference between LSTM_Acc_AVG and Val_Acc_AVG (the "difference" column in your table).

In [59]:
a3 = a3.where(a1['LSTM_Acc_AVG']>=0.53)
a3 = a3.sort_values(by='Val_Acc_AVG', ascending = False)
a3= a3.dropna()
a3

LSTM_Acc_AVG  Val_Acc_AVG  Timestep  Neurons  Dropout  Learning  \
105      0.530923     0.529236      12.0    128.0      0.1     0.001   
139      0.535701     0.528748      48.0    128.0      0.2     0.001   
180      0.530931     0.527760      96.0    128.0      0.2     0.001   
33       0.530091     0.527545       2.0    128.0      0.1     0.001   
131      0.535337     0.527382      48.0     65.0      0.2     0.001   
..            ...          ...       ...      ...      ...       ...   
129      0.530328     0.519640      48.0     65.0      0.1     0.010   
214      0.530114     0.519611     168.0     65.0      0.2     0.010   
60       0.534727     0.519477       6.0     32.0      0.3     0.010   
87       0.531014     0.519354      12.0      8.0      0.4     0.001   
94       0.531651     0.517572      12.0     32.0      0.4     0.010   

     difference  
105    0.001687  
139    0.006954  
180    0.003171  
33     0.002545  
131    0.007955  
..          ...  
129    0.010688  
214    0.010503  
60     0.015250  
87     0.011660  
94     0.014079  

[61 rows x 7 columns]

# We select this configuration:

### Timestep: 12
### Neurons: 128
### Dropout: 0.1
### Learning: 0.001
### With a difference of 0.001687132, LSTM Acc. AVG of 0.530923147, and Val Acc.AVG of 0.529236015. 